doc/dev/autograph.rst

```diff
@@ -0,0 +1,643 @@
+be assigned in **all** branches. This means that you **must** include an
+``else`` statement if you are assigning a new variable:
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     if x > 5:
+...         y = 0.4
+...     return x
+>>> f(0.5)
+AutoGraphError: Some branches did not define a value for variable 'y'
+
+If the variable previous exists before the if statement, however, this restriction
+does not apply **as long as you don't change the type**:
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     y = 0.1
+...     if x > 5:
+...         y = 0.4
+...     return y
+>>> f(0.5)
+array(0.4)
+
+If we change the type of the ``y``, however, we will need to include an
+``else`` statement to also change the type:
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     y = 0.1
+...     if x > 5:
+...         y = 4
+...     else:
+...         y = -1
+...     return y
+>>> f(0.5)
+array(-1)
+
+Compatible type assignments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Within an if statement, variable assignments must include JAX compatible
+types (Booleans, Python numeric types, JAX arrays, and PennyLane quantum
+operators). Not compatible types (such as strings) will result in an error:
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     if x > 5:
+...         y = "a"
+...     else:
+...         y = "b"
+...     return y
+>>> f(0.5)
+TypeError: Value 'a' with type <class 'str'> is not a valid JAX type
+
+
+For loops
+---------
+
+Most ``for`` loop constructs will be properly captured and compiled by AutoGraph.
+This includes automatic unpacking and enumeration through JAX arrays:
+
+>>> @qjit(autograph=True)
+... def f(weights):
+...     z = 0.
+...     for i, (x, y) in enumerate(weights):
+...         z = i * x + i ** 2 * y
+...     return z
+>>> weights = jnp.array([[0.1, 0.2, 0.3, 0.4], [0.5, 0.6, 0.7, 0.8]]).T
+>>> f(weights)
+array(8.4)
+
+This also works when looping through Python containers, **as long as the containers
+can be converted to a JAX array**:
+
+>>> weights = [[0.1, 0.2], [0.3, 0.4], [0.5, 0.6]]
+>>> f(weights)
+array(3.4)
+
+If the container cannot be converted to a JAX array (e.g., a list of strings),
+then AutoGraph will **not** capture the for loop; instead, AutoGraph will
+fallback to Python, and the loop will be unrolled at compile-time:
+
+.. code-block:: python
+
+    dev = qml.device("lightning.qubit", wires=1)
+
+    @qjit(autograph=True)
+    @qml.qnode(dev)
+    def f():
+        params = ["0", "1", "2"]
+        for x in params:
+            qml.RY(int(x) * jnp.pi / 4, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+>>> f()
+array(-0.70710678)
+
+The Python ``range`` function is also fully supported by AutoGraph, even when
+its input is a **dynamic variable** (e.g., its numeric value is only known at
+compile time):
+
+>>> @qjit(autograph=True)
+... def f(n):
+...     x = - jnp.log(n)
+...     for k in range(1, n + 1):
+...         x = x + 1 / k
+...     return x
+>>> f(100000)
+array(0.57722066)
```

Comment:
Guess the constant 😆

Reply:
![image](https://github.com/PennyLaneAI/catalyst/assets/29467667/b891e8be-5e1a-402e-a790-9456b8bf7caa)